firmware/ethereum.c

```diff
@@ -0,0 +1,551 @@
+							  const uint8_t *gas_price, uint32_t gas_price_len,
+							  const uint8_t *gas_limit, uint32_t gas_limit_len)
+{
+	bignum256 val, gas;
+	uint8_t pad_val[32];
+	char tx_value[25];
+	char gas_value[25];
+
+	memset(pad_val, 0, sizeof(pad_val));
+	memcpy(pad_val + (32 - gas_price_len), gas_price, gas_price_len);
+	bn_read_be(pad_val, &val);
+
+	memset(pad_val, 0, sizeof(pad_val));
+	memcpy(pad_val + (32 - gas_limit_len), gas_limit, gas_limit_len);
+	bn_read_be(pad_val, &gas);
+	bn_multiply(&val, &gas, &secp256k1.prime);
+
+	ethereumFormatAmount(&gas, gas_value);
+
+	memset(pad_val, 0, sizeof(pad_val));
+	memcpy(pad_val + (32 - value_len), value, value_len);
+	bn_read_be(pad_val, &val);
+
+	if (bn_is_zero(&val)) {
+		strcpy(tx_value, "message");
+	} else {
+		ethereumFormatAmount(&val, tx_value);
+	}
+
+	layoutDialogSwipe(&bmp_icon_question,
+		"Cancel",
+		"Confirm",
+		NULL,
+		"Really send",
+		tx_value,
+		"paying up to",
+		gas_value,
+		"for gas?",
+		NULL
+	);
+}
+
+/*
+ * RLP fields:
+ * - nonce (0 .. 32)
+ * - gas_price (0 .. 32)
+ * - gas_limit (0 .. 32)
+ * - to (0, 20)
+ * - value (0 .. 32)
+ * - data (0 ..)
+ */
+
+static bool ethereum_signing_check(EthereumSignTx *msg)
+{
+	if (!msg->has_nonce || !msg->has_gas_price || !msg->has_gas_limit) {
+		return false;
+	}
+	
+	if (msg->to.size != 20 && msg->to.size != 0) {
+		/* Address has wrong length */
+		return false;
+	}
+
+	// sending transaction to address 0 (contract creation) without a data field
+	if (msg->to.size == 0 && (!msg->has_data_length || msg->data_length == 0)) {
+		return false;
+	}
+
+	if (msg->gas_price.size + msg->gas_limit.size  > 30) {
+		// sanity check that fee doesn't overflow
+		return false;
+	}
+
+	return true;
+}
+
+void ethereum_signing_init(EthereumSignTx *msg, const HDNode *node)
+{
+	ethereum_signing = true;
+	sha3_256_Init(&keccak_ctx);
+
+	memset(&resp, 0, sizeof(EthereumTxRequest));
+	/* set fields to 0, to avoid conditions later */
+	if (!msg->has_value)
+		msg->value.size = 0;
+	if (!msg->has_data_initial_chunk)
+		msg->data_initial_chunk.size = 0;
+	if (!msg->has_to)
+		msg->to.size = 0;
+
+	if (msg->has_data_length) {
+		if (msg->data_length == 0) {
+			fsm_sendFailure(FailureType_Failure_Other, "Invalid data length provided");
+			ethereum_signing_abort();
+			return;
+		}
+		if (!msg->has_data_initial_chunk || msg->data_initial_chunk.size == 0) {
+			fsm_sendFailure(FailureType_Failure_Other, "Data length provided, but no initial chunk");
+			ethereum_signing_abort();
+			return;
+		}
+		/* Our encoding only supports transactions up to 2^24 bytes.  To
+		 * prevent exceeding the limit we use a stricter limit on data length.
+		 */
+		if (msg->data_length > 16000000)  {
+			fsm_sendFailure(FailureType_Failure_Other, "Data length exceeds limit");
+			ethereum_signing_abort();
+			return;
+		}
+		data_total = msg->data_length;
+	} else {
+		data_total = 0;
+	}
+	if (msg->data_initial_chunk.size > data_total) {
```

Comment:
But we want it to be executed in both cases: If data_length is missing, check that no initial chunk was given, if data_length is given, check that initial chunk is not too large.